### PR TITLE
Modal window

### PR DESCRIPTION
**Pull request by wellhairy**

### CODE REVIEW

File: assets/scss/partials/structure/_structure.scss [hunk 1]
```diff
@@ -55,6 +55,13 @@ body {
 
 .row {
   max-width: 970px;
+  position: relative;
+  &--xlarge {
```
Comment (LidorLapid) [inline assets/scss/partials/structure/_structure.scss]: These modifiers seem to be unrelated to the modal window. Can you please confirm?

File: assets/js/modules/modal.js [hunk 2]
```diff
@@ -0,0 +1,115 @@
+export default class Modal {
+  init(render) {
+    // This model window only expects string content.
```
Comment (LidorLapid) [inline assets/js/modules/modal.js]: Can we please rename all instances of 'model' to 'modal'?

File: assets/js/modules/modal.js [hunk 3]
```diff
@@ -0,0 +1,115 @@
+export default class Modal {
+  init(render) {
```
Comment (LidorLapid) [inline assets/js/modules/modal.js]: I think it would be better if we refactor this slightly. First, we can make use of a constructor function that will get automatically run when we instantiate the class. And second, I'd prefer it if the constructor/init function didn't take a parameter (or at least not this one). I think it would be better for reuse if we create a separate render function that would update the modal window after it's been created.

File: assets/js/modules/modal.js [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    // This model window only expects string content.
+    // Api's will need to be rendered else where calling
+    // this function seperatly.
+
+    if (document.querySelector('[data-modal-state]').length <= 0) {
+      return;
+    }
+
+    this.cacheContent = render;
```
Comment (LidorLapid) [inline assets/js/modules/modal.js]: This line won't be necessary once we refactor as above.

File: assets/js/modules/modal.js [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    this.buildModalWindow();
```
Comment (LidorLapid) [inline assets/js/modules/modal.js]: We could rename this function to render() and have it take the HTML string as a parameter.

File: assets/scss/partials/structure/_structure.scss (hunk 1 above)
Comment (wellhairy) [inline assets/scss/partials/structure/_structure.scss]: They originally started here to allow me to bust the model window out of our standard 970 grid primarily for desktop.

File: assets/js/modules/modal.js [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+          </div>
+        </div>
+      </div>
+    </div>`;
+
+
+    document.body.firstElementChild.insertAdjacentHTML('beforebegin', modelMarkup);
+  }
+
+  addBodyClass() {
```
Comment (LidorLapid) [inline assets/js/modules/modal.js]: Let's get rid of this function and just chuck that line into what is currently buildModalWindow(). That would make more sense as the removeModal() function takes care of the body class.

File: assets/scss/partials/structure/_structure.scss [hunk 7]
```diff
@@ -70,6 +77,50 @@ body {
...
+    content: "";
+    display: inline-block;
+    vertical-align: middle;
+    height: 100%;
+    min-height: 100vh;
+    width: 1px;
+    margin-right: -5px;
+  }
+
+  > div {
```
Comment (LidorLapid) [inline assets/scss/partials/structure/_structure.scss]: Can this be made more BEM-like? ;) Let's also think about the class name, as it wasn't obvious to me what this class did until I looked at the properties.

File: assets/scss/partials/structure/_structure.scss [hunk 8]
```diff
@@ -70,6 +77,50 @@ body {
   overflow: hidden;
 }
 
+.full-screen {
```
Comment (LidorLapid) [inline assets/scss/partials/structure/_structure.scss]: As discussed, let's refactor this slightly, e.g. rename class to .modal and get rid of the position: absolute stuff.

File: assets/scss/partials/structure/_structure.scss [hunk 9]
```diff
@@ -70,6 +77,50 @@ body {
...
+  bottom: 0px;
+  left: 0px;
+  z-index: 150;
+
+  &--fixed {
+    overflow-y: scroll;
+    position: fixed;
+  }
+
+  &--varticaly-centerd-content {
```
Comment (LidorLapid) [inline assets/scss/partials/structure/_structure.scss]: Looks like this can be removed in favour of .vertical-inline-block.

File: assets/js/modules/modal.js [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+
+  buildModalWindow() {
+    var modelMarkup =
+    `<div class="full-screen full-screen--fixed whiteBg" data-modal-window-overlay>
+      <div class="row row--xlarg">
+        <div class="column">
+          <div class="vertical-inline-block">
+            <div class="card card--specter card--big-card" data-modal-window>
+              ${this.cacheContent}
+              <a href="#" class="card--close" roll="button" aria-label="close" data-close-modal-window><span aria-hidden="true">&#10804;</span><span class="accessible-hide">Close modal window</span></a>
```
Comment (LidorLapid) [inline assets/js/modules/modal.js]: Typo: roll="button" should be role="button"

Comment (LidorLapid) [inline assets/js/modules/modal.js]: Think we should change the hidden text slightly, as most users probably wouldn't know what a modal window is. Maybe something along the lines of 'Close window' or 'Return to main content'?

File: assets/scss/partials/structure/_structure.scss (hunk 1 above)
Comment (wellhairy) [inline assets/scss/partials/structure/_structure.scss]: I've removed these from here and introduce them inside the correct branch.

File: assets/js/modules/modal.js [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+    this.modelCloseButton = document.querySelector('[data-close-modal-window]');
+  }
+
+  bindFunctions() {
+    this.focusToModalWindowBound = this.focusToModalWindow.bind(this);
+    this.closeModalClickEventBound = this.closeModalClickEvent.bind(this);
+    this.closeModalKeyUpEventBound = this.closeModalKeyUpEvent.bind(this);
+  }
+
+  attachEventListiners() {
```
Comment (LidorLapid) [inline assets/js/modules/modal.js]: Typo: "Listiners" should be "Listener" -- should be fixed across all instances

File: assets/scss/partials/elements/_modal.scss [hunk 12]
```diff
@@ -0,0 +1,7 @@
+.model-pain {
```
Comment (LidorLapid) [inline assets/scss/partials/elements/_modal.scss]: Let's rename this to .modal and chuck the .full-screen styling here instead, and then ditch the .full-screen class.

File: assets/scss/partials/structure/_cards.scss [hunk 13]
```diff
@@ -102,6 +109,30 @@
     margin-bottom: 30px;
   }
 
+  &--big-card {
```
Comment (LidorLapid) [inline assets/scss/partials/structure/_cards.scss]: Think this can be removed.

File: assets/js/modules/modal.js [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+      this.closeModal();
+    }
+  }
+
+  closeModal() {
+    this.removeEventListiners();
+    this.isOpen = false;
+    this.removeModal();
+
+    if (this.lastfocus) {
```
Comment (LidorLapid) [inline assets/js/modules/modal.js]: Shall we rename lastfocus to previouslyFocussedElement?

Comment (LidorLapid) [inline assets/js/modules/modal.js]: Under what conditions would this conditional evaluate to a falsey value?

File: assets/js/modules/modal.js [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+      this.closeModal();
+    }
+
+    if(e.target !== this.modalWindow && !this.modalWindow.contains(e.target)) {
+      this.closeModal();
+    }
+  }
+
+  closeModalKeyUpEvent(e) {
+    if (!e.keyCode || e.keyCode === 27) {
```
Comment (LidorLapid) [inline assets/js/modules/modal.js]: Under what conditions would e.keyCode be falsey?

File: assets/js/modules/modal.js [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+  cacheElements() {
+    this.lastFocus = document.activeElement;
+    this.isOpen = true;
+    this.modelOverlay = document.querySelector('[data-modal-window-overlay]');
+    this.modalWindow = document.querySelector('[data-modal-window]');
+    this.modelOpenButton = document.querySelector('[data-open-modal-window]');
+    this.modelCloseButton = document.querySelector('[data-close-modal-window]');
+  }
+
+  bindFunctions() {
```
Comment (LidorLapid) [inline assets/js/modules/modal.js]: Change as discussed.

File: assets/js/modules/modal.js [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+
+
+    document.body.firstElementChild.insertAdjacentHTML('beforebegin', modelMarkup);
+  }
+
+  addBodyClass() {
+    document.querySelector('body').classList.add('overflow-hidden');
+  }
+
+  setOpenFocusState() {
```
Comment (LidorLapid) [inline assets/js/modules/modal.js]: Change as discussed.

File: assets/js/modules/modal.js (hunk 14 above)
Comment (wellhairy) [inline assets/js/modules/modal.js]: Let me look into this a little. I think it's in the use case that there was no  focused element like in the case of an exit survey - opening before you've navigated the page.

File: assets/js/modules/modal.js (hunk 15 above)
Comment (wellhairy) [inline assets/js/modules/modal.js]: I really can't remeber what the `!e.keyCode ||` section of code is used for

File: assets/js/modules/modal.js (hunk 6 above)
Comment (wellhairy) [inline assets/js/modules/modal.js]: Done

File: assets/scss/partials/structure/_structure.scss (hunk 9 above)
Comment (wellhairy) [inline assets/scss/partials/structure/_structure.scss]: Agree, this has been removed

File: assets/js/modules/modal.js (hunk 10 above)
Comment (wellhairy) [inline assets/js/modules/modal.js]: Thanks, done

Comment (wellhairy) [inline assets/js/modules/modal.js]: Done

File: assets/js/modules/modal.js (hunk 11 above)
Comment (wellhairy) [inline assets/js/modules/modal.js]: Thanks, done

File: assets/scss/partials/elements/_modal.scss (hunk 12 above)
Comment (wellhairy) [inline assets/scss/partials/elements/_modal.scss]: I've wrapped all this up one way or another

File: assets/scss/partials/structure/_cards.scss (hunk 13 above)
Comment (wellhairy) [inline assets/scss/partials/structure/_cards.scss]: Moved under modal class

File: assets/js/modules/modal.js (hunk 14 above)
Comment (wellhairy) [inline assets/js/modules/modal.js]: We can I guess

File: assets/js/modules/modal.js (hunk 2 above)
Comment (wellhairy) [inline assets/js/modules/modal.js]: Done

File: assets/scss/partials/structure/_structure.scss (hunk 7 above)
Comment (wellhairy) [inline assets/scss/partials/structure/_structure.scss]: We need to talk this comment over, it'd be nice to trya nd define the class name "vertical-inline-block" together

File: assets/js/modules/modal.js [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+  removeModal() {
+    document.querySelector('body').classList.remove('overflow-hidden');
+    this.modelOverlay.parentNode.removeChild(this.modelOverlay);
+  }
+
+  buildModal(htmlString) {
+    // This modal window only expects string content.
+    // Api's will need to be rendered else where calling
+    // this function seperatly.
+    var modelMarkup =
```
Comment (natmunckton) [inline assets/js/modules/modal.js]: Need consistent spelling of modal throughout
